app/moderation-algo/main.py

```diff
@@ -0,0 +1,204 @@
+    REDIS_PORT=6379
+    REDIS_USERNAME='default'
+    REDIS_PASSWORD='redispassword'
+    return redis.StrictRedis(host=REDIS_HOST, port=REDIS_PORT, username=REDIS_USERNAME, password=REDIS_PASSWORD,
+                             charset='utf-8', decode_responses=True)
+
+@app.route('/', methods=['POST'])
+def HandleUserInput():
+    # Check if the request contains JSON data
+    if request.is_json:
+        # NOTE: Make sure Google API is initialized at this point
+        model = 'gemini-pro'
+
+        # Connect to Redis with a week expiration time
+        secInAWeek = 604800
+        r = ConnectToRedis()
+
+        # Decode which stage the request is going to
+        stage = request.get_json().get('stage')
+
+        # Stage 1: Extract the overall issue and the topics and definitions from reading materials
+        if(stage == 'extraction'):
+            # Extract and issue and topics for later use
+            reading_materials = request.get_json().get('reading_materials')
+            if(not reading_materials):
+                return jsonify({'ERROR': 'Missing field "reading_materials" in request data'}), 422 # HTTP unprocessable Entity
+            issue = ExtractIssueFromReadingMaterials(model, reading_materials)
+            topics = ExtractTopicsDescriptions(model, reading_materials)
+
+            # Save them for later use, expiring in a week
+            r.set('moderation_issue', issue, ex=secInAWeek)
+            r.set('moderation_topics', json.dumps(topics), ex=secInAWeek)
+            r.set('moderation_reading_materials', reading_materials, ex=secInAWeek)
+
+            # Construct and return the response
+            response = {
+                'issue': issue,
+                'topics': topics
+            }
+            return jsonify(response), 200 # HTTP success
+
+        # Stage 2: Interactively finalize the set of topics and definitions.
+        elif(stage == 'interactive'):
+            # At this point, the extracted topics and list of locked topics should already be available in Redis
+            # If lockedTopics does not exist, create a new empty list for it
+            lockedTopics = r.get('moderation_lockedTopics')
+            if(lockedTopics):
+                lockedTopics = json.loads(lockedTopics)
+            else:
+                lockedTopics = []
+                r.set('moderation_lockedTopics', json.dumps(lockedTopics))
+            topics = json.loads(r.get('moderation_topics'))
+            if(not topics):
+                return jsonify({'ERROR': 'Unable to find value(s) in stored data. Please rerun Stages 1 and 2.'}), 400 # HTTP bad request
+            
+            # Perform the user designated action
+            action = request.get_json().get('action')
+            if(not action):
+                return jsonify({'ERROR': 'Missing field "action" in request data'}), 422 # HTTP unprocessable Entity
+            action = action.lower()
+            
+            if(action == 'lock'):
```

Comment:
It would be nice for many of the actions to have an alternate version that supports a list of topics rather than just a single one. I think it would be useful mainly for the lock, unlock, add, and remove actions. Would allow for minimizing the api calls and allow the user to do multiple edits on the client at once rather than one at a time.